File: ruby-on-rails/app/models/algorithms/fine_automata.rb
class FineAutomata
  def initialize(no_of_chars)
    @no_of_chars = no_of_chars
  end

  def get_next_state(pat, m, state, x)
    if state < m && x == pat[state].ord
      return state + 1
    end
    ns = state
    while ns > 0
      if pat[ns - 1].ord == x
        i = 0
        while i < ns - 1
          if pat[i] != pat[state - ns + 1 + i]
            break
          end
          i += 1
        end
        if i == ns - 1
          return ns
        end
      end
      ns -= 1
    end
    0
  end

  def compute_tf(pat, m, tf)
    state = 0
    while state <= m
      x = 0
      while x < @no_of_chars
        tf[state][x] = self.get_next_state(pat, m, state, x)
        x += 1
      end
      state += 1
    end
  end

  def search(pat, txt)
    _m = pat.length
    _n = txt.length
    _tf = Array.new(_m + 1) { Array.new(@no_of_chars) }
    self.compute_tf(pat, _m, _tf)
    state = 0
    i = 0
    result_list = []
    while i < _n
      state = _tf[state][txt[i].ord]
      if state == _m
        result_list.push("Pattern found at index " + (i - _m + 1).to_s)
      end
      i += 1
    end
    result_list
  end

=begin
  def self.main
    pat = "AABAACAADAABAAABAA".chars
    txt = "AABA".chars
    self.search(txt, pat)
  end
=end

end

File: spring/src/main/kotlin/net/jackowski/spring/util/algorithms/FineAutomata.kt
package net.jackowski.spring.util.algorithms


class FineAutomata {

    companion object {
        private var NO_OF_CHARS = 256
    }

    private fun getNextState(
        pat: String, m: Int,
        state: Int, x: Int
    ): Int {

        // If the character c is same as next
        // character in pattern,then simply
        // increment state
        if (state < m && x == pat[state].code) return state + 1

        // ns stores the result which is next state
        var i: Int

        // ns finally contains the longest prefix
        // which is also suffix in "pat[0..state-1]c"

        // Start from the largest possible value
        // and stop when you find a prefix which
        // is also suffix
        var ns: Int = state
        while (ns > 0) {
            if (pat[ns - 1].code == x) {
                i = 0
                while (i < ns - 1) {
                    if (pat[i] != pat[state - ns + 1 + i]) break
                    i++
                }
                if (i == ns - 1) return ns
            }
            ns--
        }
        return 0
    }

    /* This function builds the TF table which
	represents Finite Automata for a given pattern */
    private fun computeTF(pat: String, m: Int, tf: Array<IntArray>) {
        var x: Int
        var state = 0
        while (state <= m) {
            x = 0
            while (x < NO_OF_CHARS) {
                tf[state][x] = getNextState(pat, m, state, x)
                x++
            }
            state++
        }
    }

    /* Prints all occurrences of pat in txt */
    fun search(pat: String, txt: String): List<String> {
        val m = pat.length
        val n = txt.length
        val tf = Array(m + 1) { IntArray(NO_OF_CHARS) }
        computeTF(pat, m, tf)

        // Process txt over FA.
        var state = 0
        var i = 0
        val resultList: MutableList<String> = ArrayList()
        while (i < n) {
            state = tf[state][txt[i].code]
            if (state == m) resultList.add("Pattern found at index ${i - m + 1}")
            i++
        }
        return resultList
    }
}

// Driver code
/*
fun main() {
    val pat = "AABAACAADAABAAABAA"
    val txt = "AABA"
    val f = FineAutomata()
    f.search(txt, pat)
}*/
